.github/scripts/userbenchmark/aicluster.py

```diff
@@ -0,0 +1,200 @@
+        assert filename, f"Expected non-empty filename from key {key}."
+        with open(os.path.join(dest_dir, filename), 'wb') as f:
+            self.s3.download_fileobj(self.bucket, key, f)
+
+    def upload_file(self, prefix, file_path):
+        file_name = file_path.name
+        s3_key = f"{self.object}/{prefix}/{file_name}" if prefix else f"{self.object}/{file_name}"
+        response = self.s3.upload_file(str(file_path), self.bucket, s3_key)
+        print(f"S3 client response: {response}")
+
+    def exists(self, prefix, file_name):
+        """Test if the key object/prefix/file_name exists in the S3 bucket.
+           If True, return the S3 object key. Return None otherwise. """
+        s3_key = f"{self.object}/{prefix}/{file_name}" if prefix else f"{self.object}/{file_name}"
+        result = self.s3.list_objects_v2(Bucket=self.bucket, Prefix=s3_key)
+        if 'Contents' in result:
+            return s3_key
+        return None
+
+    def list_directory(self, directory=None):
+        """List the directory files in the S3 bucket path.
+           If the directory doesn't exist, report an error. """
+        prefix = f"{self.object}/{directory}/" if directory else f"{self.object}/"
+        pages = self.s3.get_paginator("list_objects").paginate(Bucket=self.bucket, Prefix=prefix)
+        keys = filter(lambda x: not x == prefix, [e['Key'] for p in pages for e in p['Contents']])
+        return list(keys)
+    
+    def get_filename_from_key(object_key):
+        filename = object_key.split('/')[-1]
+        return filename
+
+def determine_success_today(index, allow_yesterday=True):
+    """
+    Determine whether today or yesterday's run is successful.
+    """
+    # get today's date in UTC
+    today = datetime.datetime.utcnow().date()
+    today_str = f"metrics-{today.strftime('%Y%m%d')}"
+    yesterday = (today - datetime.timedelta(days=1))
+    yesterday_str = f"metrics-{yesterday.strftime('%Y%m%d')}"
+    for index_key in index:
+        # check if today or yesterday's date exists in the index
+        if today_str in index_key:
+            print(f"Found today run log: {index_key} ")
+            return True
+        if allow_yesterday and yesterday_str in index_key:
+            print(f"Found yesterday run log: {index_key} ")
+            return True
+    # not found, the last run probably failed
+    return False
+
+def get_metrics_index(s3, benchmark_name, work_dir):
+    """
+    1. Try to download the index file from S3, 
+    2. if not found, create an initial one with the metrics files from S3 directory
+    3. Otherwise, compare the downloaded index file with the metrics file list, update the index file, and return
+    """
+    def gen_index_obj(index_key):
+        "download and load the index file if exists, otherwise, return empty object."
+        if not index_key:
+            return {}
+        filename = S3Client.get_filename_from_key(index_key)
+        s3.download_file(index_key, work_dir)
```

Comment:
do we need retry if say, s3 failed? Or will AWS automatically do that for us?

Reply:
Usually S3 is reliable, if it fails probably retry won't help, so we will directly fail he entire workflow, and look into this manually.